playpen.py

```diff
@@ -1,8 +1,63 @@
 #!/usr/bin/env python3
 
 import subprocess
+import shlex #for quote() - python3 required
+import re #for re.match()
 
-def execute(version, command, arguments, data=None):
+def execute(version, command, arguments, data=None, env_vars=None):
+    #Note: arguments and env_vars are expected to be of tuple type, not list,
```

Comment:
**tl;dr4**: ok, don't read any of this, you'll just see what I mean in the (future)commit - and if you really want me to avoid escaping args(which is done ONLY when backtrace is on) then, fine :)

**tl;dr3:** I can reduce this logic as you want it, but to ensure consistency of subprocess execution between when backtrace is on and when it's off, I have to quote the `arguments` passed to `command` but only in the former case(of backtrace being on). Else, they are still consistent(for our known args thus far) but only until any one arg(added by someone in the future) contains a bad char(space, quotes, backticks, dollar signs, etc.)

**tl;dr2:** the way we decided to pass the env. var RUST_BACKTRACE(via `dash -c "export RUST_BACKTRACE=1; command args"`), requires to escape/quote the args passed to `command`(eg. to `evaluate.sh`) in order to maintain the same logic for when backtrace is off(ie. executing just `command args`, where the args are each elements of a tuple when passed to `subprocess.Popen`, whereas with the `dash -c` logic the args and command are both just one element of a tuple like so: `("/usr/bin/dash", "-c", "command args")) - ok i'm bad at tl;dr-s

**tl;dr:** Should I escape/quote the args when backtrace is on? or else the logic for passing args to command but-only-when-backtrace-is-on is in danger of breaking-but-only-in-the-case-when future new args would contain any spaces, dollar signs, single or double quotes, backticks etc.

**I can definitely reduce this as you want it**, but one thing I still have to do is to quote the args(but only in the case when backtrace is on, because this is the only time when we wrap the `command` within the `dash` shell), or else `dash` won't have the same guarantee that just running the `command`(without `dash` wrapped around it, that is) *currently* has, which is: an arg for example can contain spaces, and it would be seen as one arg not two; to ensure this is true even with `dash -c "command args"` I must escape the args (even though `command args` is sent as one argument to the `dash` command, it's the `command` that needs to see each of the args as whole as they are like when sent directly through `command args` without the `dash` wrapped around)

For example:
`evaluate.sh "this is one arg"`
vs.
`dash "-c" "evaluate.sh this is one arg"`
(in the above, I added double quotes for visibility only! to signify the arg boundaries)
So `this is one arg` becomes 4 args in the latter case! Unless I escape it (with shlex.quote()), in which case it becomes:
`dash "-c" "evaluate.sh 'this is one arg'"`
(with single quotes present in string; to reiterate, the double quotes are not present in string - they're just for visibility here)

You could argue that since we do know what the args are right NOW in this moment of time(`--test`, `--color=always`, `-C`, `opt-level=0`, `-g`, maybe others), we do not need to escape(quote) any! But who would remember to look in web.py's `execute` function (which is where I'll move the overly simplified specific logic for this - I have to, because it's no longer generic to stay here in `playpen.py`) and to make sure the supposedly newly passed args(of the future) are not split into two args because they now contain spaces(or backticks or quotes, or dollar signs).